File: sources/scalac/util/NameTransformer.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
**                                                                      **
** $Id$
\*                                                                      */

package scalac.util;

/** A name transformer for replacing operator symbols in names by predefined
 *  names of the form $opname.
 *
 *  @author     Martin Odersky, Christine Roeckl
 *  @version    1.1
 */
public class NameTransformer {

    public static String[] operatorName = new String[128];

    static {
       operatorName['~'] = "$tilde";
       operatorName['='] = "$eq";
       operatorName['<'] = "$less";
       operatorName['>'] = "$greater";
       operatorName['!'] = "$bang";
       operatorName['#'] = "$hash";
       operatorName['%'] = "$percent";
       operatorName['^'] = "$up";
       operatorName['&'] = "$amp";
       operatorName['|'] = "$bar";
       operatorName['*'] = "$times";
       operatorName['/'] = "$div";
       operatorName['+'] = "$plus";
       operatorName['-'] = "$minus";
       operatorName[':'] = "$colon";
    }

    /** Replace operator symbols by corresponding "$op_name" in names.
     */
    public static Name encode(Name name) {
       int i = name.index;
       int len = i + name.length();
       StringBuffer res = new StringBuffer();
       while (i < len) {
         byte c = Name.names[i++];
         String nop = operatorName[c];
         if (nop == null)
             res.append((char)c);
         else
             res.append(nop);
       }
       return Name.fromString(res.toString());
    }

    /** Replace "$op_name" by corresponding operator symbols in names.
     */
    public static Name decode(Name name) {
        int i = name.index;
	int length = i + name.length();
	StringBuffer res = new StringBuffer();
	while (i < length) {
	    byte c = Name.names[i++];
	    if ((char)c == '$') {
		String op = cutOut(i, length);
		res.append(decode("$" + op));
		i = i + op.length();
	    }
	    else res.append((char)c);
	}
       return Name.fromString(res.toString());
    }

    /** Decodes (a prefix of) a string.
     */
    public static String decode(String string) {
	for (int i = string.length(); i > 0; i--) {
	    String prefix = string.substring(0, i);
	    String c = lookup(prefix);
	    if (c != null) {
		String suffix = string.substring(i, string.length());
		return c + suffix;
	    }
	}
	return string;
    }

    /** Cuts out the name of an operator plus succeeding characters.
     *  Does NOT return the preceeding '$' symbol.
     */
    private static String cutOut(int pos, int length) {
	int i = pos;    // avoid side-effects
	StringBuffer res = new StringBuffer();
	while (i < length) {
	    char c = (char)Name.names[i++];
	    if (c == '$') return res.toString();
	    else res.append(c);
	}
	return res.toString();
    }

    /** Looks up the array entry for the operator name.
     */
    private static String lookup(String string) {
	for (int i = 0; i < 128; i++) {
	    if ((operatorName[i] != null) && (string.compareTo(operatorName[i]) == 0))
		return String.valueOf((char)i);
	}
	return null;
    }
}

File: src/compiler/scala/tools/nsc/util/NameTransformer.scala
/* NSC -- new Scala compiler
 * Copyright 2005-2006 LAMP/EPFL
 * @author  Martin Odersky
 */
// $Id$

package scala.tools.nsc.util

import compat.StringBuilder

object NameTransformer {
  private val nops = 128
  private val ncodes = 26 * 26

  private class OpCodes(val op: char, val code: String, val next: OpCodes)

  private val op2code = new Array[String](nops)
  private val code2op = new Array[OpCodes](ncodes)

  private def enterOp(op: char, code: String) = {
    op2code(op) = code
    val c = (code.charAt(1) - 'a') * 26 + code.charAt(2) - 'a'
    code2op(c) = new OpCodes(op, code, code2op(c))
  }

  enterOp('~', "$tilde")
  enterOp('=', "$eq")
  enterOp('<', "$less")
  enterOp('>', "$greater")
  enterOp('!', "$bang")
  enterOp('#', "$hash")
  enterOp('%', "$percent")
  enterOp('^', "$up")
  enterOp('&', "$amp")
  enterOp('|', "$bar")
  enterOp('*', "$times")
  enterOp('/', "$div")
  enterOp('+', "$plus")
  enterOp('-', "$minus")
  enterOp(':', "$colon")
  enterOp('\\', "$bslash")
  enterOp('?', "$qmark")
  enterOp('@', "$at")

  /** Replace operator symbols by corresponding "<code>$op_name</code>".
   *
   *  @param name ...
   *  @return     ...
   */
  def encode(name: String): String = {
    var buf: StringBuilder = null
    val len = name.length()
    var i = 0
    while (i < len) {
      val c = name charAt i
      if (c < nops && (op2code(c) ne null)) {
        if (buf eq null) {
          buf = new StringBuilder()
          buf.append(name.substring(0, i))
        }
        buf.append(op2code(c))
      } else if (buf ne null) {
        buf.append(c)
      }
      i = i + 1
    }
    if (buf eq null) name else buf.toString()
  }

  /** Replace <code>$op_name</code> by corresponding operator symbol.
   *
   *  @param name0 ...
   *  @return      ...
   */
  def decode(name0: String): String = {
    //System.out.println("decode: " + name);//DEBUG
    val name = if (name0.endsWith("<init>")) name0.substring(0, name0.length() - ("<init>").length()) + "this"
               else name0;
    var buf: StringBuilder = null
    val len = name.length()
    var i = 0
    while (i < len) {
      var ops: OpCodes = null
      val c = name charAt i
      if (c == '$' && i + 2 < len) {
        val ch1 = name.charAt(i+1)
        if ('a' <= ch1 && ch1 <= 'z') {
          val ch2 = name.charAt(i+2)
          if ('a' <= ch2 && ch2 <= 'z') {
            ops = code2op((ch1 - 'a') * 26 + ch2 - 'a')
            while ((ops ne null) && !name.startsWith(ops.code, i)) ops = ops.next;
            if (ops ne null) {
              if (buf eq null) {
                buf = new StringBuilder()
                buf.append(name.substring(0, i))
              }
              buf.append(ops.op)
              i = i + ops.code.length()
            }
          }
        }
      }
      if (ops eq null) {
        if (buf ne null) buf.append(c)
        i = i + 1
      }
    }
    //System.out.println("= " + (if (buf == null) name else buf.toString()));//DEBUG
    if (buf eq null) name else buf.toString()
  }
}
